src/MemStore.cc

```diff
@@ -492,10 +497,26 @@ MemStore::copyFromShm(StoreEntry &e, const sfileno index, const Ipc::StoreMapAnc
             const StoreIOBuffer sliceBuf(wasSize - prefixSize,
                                          e.mem_obj->endOffset(),
                                          page + prefixSize);
-            if (!copyFromShmSlice(e, sliceBuf, wasEof))
-                return false;
+
+            copyFromShmSlice(e, sliceBuf);
             debugs(20, 8, "entry " << index << " copied slice " << sid <<
                    " from " << extra.page << '+' << prefixSize);
+
+            // parse headers if needed; they might span multiple slices!
+            if (rep->pstate < Http::Message::psParsed) {
+                // XXX: have to copy because httpMsgParseStep() requires 0-termination
+                replyBuffer.append(sliceBuf.data, sliceBuf.length);
+                replyBuffer.terminate();
+                auto error = Http::scNone;
+
+                if (rep->parse(replyBuffer.buf, replyBuffer.size, wasEof, &error)) {
+                    assert(rep->pstate == Http::Message::psParsed);
+                } else if (error) {
+                    return false;
```

Comment:
It feels like we should move the old "Corrupted mem-cached headers" debugs() to here because inability to parse packed headers sounds like a Squid BUG. I would even add a `BUG: ` prefix to that line. Or did I miss an important reason for dropping that debugs()?

Reply:
Moved.
I could not remember why I removed this debugging. Probably I assumed that Http::Message::parse() reports about 'too large header' problem itself. However, there are many other possible parsing problems in Http::Message::parse(), and we should report about them.